### PR TITLE
python3Packages.skein: init at 0.8.0

### PR DESCRIPTION
**Pull request by alexbiehl**
###### Motivation for this change

This PR adds the skein package to python3Modules. The build process is a bit hairy as it depends on an integrated Maven build from within the setup.py. I pulled it out and build the JAR separately and made the fetching of the Java dependencies into a fixed output derivation.

The Maven build downloads some native executables for dealing with ProtocolBuffer generation and thus needs some patchelf magic which further complicates the build. 

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by shlevy — APPROVED** [review]

File: pkgs/development/python-modules/skein/default.nix [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+    buildPhase = ''
+      mvn --offline -f java/pom.xml package -Dmaven.repo.local="${skeinRepo}" -Dskein.version=${version} -Dversion=${version}
+    '';
+
+    installPhase = ''
+      # Making sure skein.jar exists skips the maven build in setup.py
+      mv java/target/skein-*.jar $out
+    '';
+  };
```
Comment (jonringer) [inline pkgs/development/python-modules/skein/default.nix:85]: I would move these into separate files:
```suggestion
let
  src = fetchPypi {... };
  skeinRepo = callPackage ./skeinrepo.nix { inherit src; };
  skeinJar = callPackage ./skeinjar.nix { inherit src; };
in
buildPythonPackage rec {
...
```

File: pkgs/development/python-modules/skein/default.nix [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  preBuild = ''
+    # Making sure skein.jar exists skips the maven build in setup.py
+    mkdir -p skein/java
+    ln -s ${skeinJar} skein/java/skein.jar
+  '';
+
+  meta = with stdenv.lib; {
+    homepage = "https://jcristharif.com/skein";
+    description = "A tool and library for easily deploying applications on Apache YARN";
+    license = licenses.bsd3;
```
Comment (jonringer) [inline pkgs/development/python-modules/skein/default.nix:98]: please add yourself as a maintainer

File: pkgs/development/python-modules/skein/default.nix [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  };
+
+  skeinRepo = stdenv.mkDerivation rec {
+    name = "${pname}-${version}-maven-repo";
+
+    inherit src;
+
+    nativeBuildInputs = [ maven ] ++ lib.optional stdenv.isLinux autoPatchelfHook;
+
+    buildPhase = "";
```
Comment (jonringer) [inline pkgs/development/python-modules/skein/default.nix:29]:
```suggestion
    doBuild = false;
```

File: pkgs/development/python-modules/skein/default.nix [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+      rm $(find $out -name _remote.repositories)
+      rm $(find $out -name resolver-status.properties)
+    '';
+
+    outputHashMode = "recursive";
+    outputHashAlgo = "sha256";
+    outputHash = if stdenv.isLinux
+      then "12f0q3444qw6y4f6qsa9540a0fz4cgi844zzi8z1phqn3k4dnl6v"
+      else "0bjbwiv17cary1isxca0m2hsvgs1i5fh18z247h1hky73lnhbrz8";
+
```
Comment (jonringer) [inline pkgs/development/python-modules/skein/default.nix:64]: is this trying to make it a fixed output? seems odd with mkDerivation

Comment (shlevy) [reply pkgs/development/python-modules/skein/default.nix:64]: This is how fixed output derivations are implemented under the hood. There's no existing fetchFoo that will do this right, but it is still a normal fixed output approach.

Comment (jonringer) [reply pkgs/development/python-modules/skein/default.nix:64]: interesting, but I still don't see the value in defining the outputHash if nix is able to determine a derivation hash

Comment (alexbiehl) [reply pkgs/development/python-modules/skein/default.nix:64]: > I still don't see the value in defining the outputHash if nix is able to determine a derivation hash

The maven dependencies are not managed by nix. IMHO this derivation is really like a complex `fetchurl` which downloads stuff from the internet. I wouldn't want to use a fetchurl without an expected hash too. 

There are some examples in nixpkgs of the like. Most notably buildRustPackage which fetches dependencies from upstream and expects the user to provide hash to guarantee reproducibility.

Comment (shlevy) [reply pkgs/development/python-modules/skein/default.nix]: @jonringer This drv fetches from the net. It needs to be fixed output.

File: pkgs/top-level/python-packages.nix [hunk 5]
```diff
@@ -6338,6 +6338,8 @@ in {
 
   python_statsd = callPackage ../development/python-modules/python_statsd { };
 
+  skein = disabledIf isPy27 (callPackage ../development/python-modules/skein { });
```
Comment (jonringer) [inline pkgs/top-level/python-packages.nix:6341]: buildPythonPackage can handle this in the expression
```suggestion
  skein = callPackage ../development/python-modules/skein { };
```

File: pkgs/development/python-modules/skein/default.nix [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  pname = "skein";
+  version = "0.8.0";
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "0nb64p1hzshgi1kfc2jx1v9vn8b0wzs50460wfra3fsxh0ap66ab";
+  };
+  skeinJar = callPackage ./skeinjar.nix { inherit src version; };
+in
+buildPythonPackage rec {
+  inherit pname version src;
```
Comment (jonringer) [inline pkgs/development/python-modules/skein/default.nix:19]:
```suggestion
  inherit pname version src;
  disabled = isPy27;
```

File: pkgs/development/python-modules/skein/default.nix [hunk 7]
```diff
@@ -0,0 +1,36 @@
+{ buildPythonPackage
+, callPackage
+, fetchPypi
```
Comment (jonringer) [inline pkgs/development/python-modules/skein/default.nix:3]:
```suggestion
, fetchPypi
, isPy27
```

File: pkgs/development/python-modules/skein/default.nix [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+, jre
+, lib
+, pythonPackages
+, stdenv
+}:
```
Comment (jonringer) [inline pkgs/development/python-modules/skein/default.nix:8]:
```suggestion
}:

```

**Comment by jonringer** [comment]
cc @FRidh

**Comment by alexbiehl** [comment]
Thank you for your suggestions @jonringer! All of them applied!

**Comment by alexbiehl** [comment]
Gentle ping for another round of reviews @jonringer @FRidh

**Review by imalsogreg — APPROVED** [review]
Looks good to me! I'd love to use this at work, any timeline for merge?

**Pull request merged**